File: 算法/算法导论代码/第零章 一些常见算法/quick_sort.c
#include <stdio.h>
#include <stdlib.h>

void quickSort(int *arr, int start, int end);
int partiton(int *arr, int start, int end);

int main(int argc, char const *argv[])
{
	int a[10] = {5, 9, 4, 7, 6, 3, 8, 10, 2, 1};
	quickSort(a, 0, 9);
	printf("final result:\n");
	for (int i = 0; i < 10; ++i)
	{
		printf("%d ", a[i]);
	}
	printf("\n");
	return 0;
}

/**
 * 快速排序
 * @param arr   [description]
 * @param start [description]
 * @param end   [description]
 */
void quickSort(int *arr, int start, int end)
{
	if (start == end)
	{
		return;
	}

	int index = partiton(arr, start, end);
	printf("index position = %d\n", index);
	printf("current array: ");
	for (int i = 0; i < 10; ++i)
	{
		printf("%d ", arr[i]);
	}
	printf("\n");
	if (index > start)
	{
		quickSort(arr, start, index - 1);
	}
	if (index < end)
	{
		quickSort(arr, index + 1, end);
	}
}

int partiton(int *arr, int start, int end)
{
	if (start == end)
	{
		return start;
	}
	int pivot = arr[start];
	int i = start;
	int j = end;
	while(i < j)
	{
		while(i < j && arr[j] > pivot)
		{
			j--;
		}
		while(i < j && arr[i] <= pivot)
		{
			i++;
		}
		int temp = arr[i];
		arr[i] = arr[j];
		arr[j] = temp;
	}
	int temp = arr[i];
	arr[i] = arr[start];
	arr[start] = temp;
	return i;
}

File: 算法/算法导论代码/第零章 一些常见算法/quick_sort.py

def partition(arr, start, end):
    if start == end:
        return start
    i = start
    j = end
    pivot = arr[start]

    while i < j:
        while i < j and arr[j] > pivot:
            j = j - 1
        while i < j and arr[i] <= pivot:
            i = i + 1
        temp = arr[i]
        arr[i] = arr[j]
        arr[j] = temp

    temp = arr[start]
    arr[start] = arr[i]
    arr[i] = temp
    return i


def quick_sort(arr, start, end):
    print(arr)
    if start == end:
        return
    index = partition(arr, start, end)
    if index > start:
        quick_sort(arr, start, index - 1)
    if index < end:
        quick_sort(arr, index + 1, end)

if __name__ == '__main__':

	t = [3, 100, 56, 32, 11, 9, 5, 2, 999, 13]
    quick_sort(t, 0, len(t) - 1)
    print(t)